### ISSUE TITLE
Improved CLI Argument Parsing

### ISSUE THREAD

**Issue by zicklag** [gfx-rs/naga#1107]
Would you be open to a PR for improving the current Naga CLI argument parsing?

Nowadays I tend to prefer [structopt](https://docs.rs/structopt/0.3.22/structopt/), but if you were concerned about code size or anything I've round [argh](https://docs.rs/argh/0.1.5/argh/) a good minimal alternative.

**Comment by kvark** [comment]
I was looking at https://github.com/blyxxyz/lexopt recently, and I couldn't convince myself if there is an improvement to be offered on top of what we have (very basic parsing). Would it support more syntax that we don't have? Or produce better error messages?

**Comment by zicklag** [comment]
The first thing I've wanted was nicer help generation and error handling. For instance, right now when I type `naga -h` it says:

```
thread 'main' panicked at 'Input has no extension?', bin/convert.rs:100:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Also, the help message is not synchronized with the actual options:

```
$ naga --help
Call with <input> <output> [<options>]
```

I didn't know until just now, looking at the source code that there were so many options. By using `argh`, we get this instead:

```
$ naga -h
Unrecognized argument: -h
$ naga --help
Usage: naga <input> [<outputs...>] [--validate] [--index-bounds-check-policy <index-bounds-check-policy>] [--flow-dir <flow-dir>] [--entry-point <entry-point>] [--profile <profile>] [--shader-model <shader-model>]

Translate shaders to different formats

Options:
  --validate        validate the shader during translation
  --index-bounds-check-policy
                    what policy to use for index bounds checking. May be
                    `Restrict`, `ReadZeroSkipWrite`, or `UndefinedBehavior`
  --flow-dir        directory to dump the SPIR-V flow dump to
  --entry-point     the shader entrypoint to use when compiling to GLSL
  --profile         the shader profile to use, for example `es 330` or `core
                    400`, if translating to GLSL
  --shader-model    the shader model to use if targeting HSLS May be `50`, 51`,
                    or `60`
  --help            display usage information
```

Also, it doesn't seem like there's really any extra compile time ( comparing what cargo had to do after I added the dependency incrementally ) because we already include the proc macro dependencies so those don't push up the build time at all.

And here's the entirety of the argument parsing code:

```rust
/// Translate shaders to different formats
#[derive(argh::FromArgs, Debug, Clone)]
struct Args {
    /// validate the shader during translation
    #[argh(switch)]
    validate: bool,

    /// what policy to use for index bounds checking.
    ///
    /// May be `Restrict`, `ReadZeroSkipWrite`, or `UndefinedBehavior`
    #[argh(option)]
    index_bounds_check_policy: Option<IndexBoundsCheckPolicyArg>,

    /// directory to dump the SPIR-V flow dump to
    #[argh(option)]
    flow_dir: Option<String>,

    /// the shader entrypoint to use when compiling to GLSL
    #[argh(option)]
    entry_point: Option<String>,

    /// the shader profile to use, for example `es 330` or `core 400`, if translating to GLSL
    #[argh(option)]
    profile: Option<String>,

    /// the shader model to use if targeting HSLS
    ///
    /// May be `50`, 51`, or `60`
    #[argh(option)]
    shader_model: Option<ShaderModelArg>,

    /// the input file
    #[argh(positional)]
    input: String,

    /// the output file
    #[argh(positional)]
    outputs: Vec<String>,
}

/// Newtype so we can implement [`FromStr`] for [`IndexBoundsCheckPolicy`].
#[derive(Debug, Clone)]
struct IndexBoundsCheckPolicyArg(IndexBoundsCheckPolicy);

impl FromStr for IndexBoundsCheckPolicyArg {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(match s.to_lowercase().as_str() {
            "restrict" => IndexBoundsCheckPolicy::Restrict,
            "readzeroskipwrite" => IndexBoundsCheckPolicy::ReadZeroSkipWrite,
            "undefinedbehavior" => IndexBoundsCheckPolicy::UndefinedBehavior,
            _ => return Err("Invalid value for --index-bounds-check-policy".into()),
        }))
    }
}

/// Newtype so we can implement [`FromStr`] for [`ShaderModel`].
#[derive(Debug, Clone)]
struct ShaderModelArg(ShaderModel);

impl FromStr for ShaderModelArg {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(match s.to_lowercase().as_str() {
            "50" => ShaderModel::V5_0,
            "51" => ShaderModel::V5_1,
            "60" => ShaderModel::V6_0,
            _ => return Err("Invalid value for --index-bounds-check-policy".into()),
        }))
    }
}
```

**Comment by kvark** [comment]
Ok, that seems useful. Thank you for an extensive answer!

**Comment by Gordon-F** [comment]
@zicklag Do you have any information about how much the compilation time increases with `argh`?

**Comment by zicklag** [comment]
I just did a clean build with `-Z timings`:

![image](https://user-images.githubusercontent.com/25393315/126345606-2137275c-9714-4a96-a036-9462745e58ba.png)


crate  | compile | codegen
 -- | -- | -- 
argh_derive v0.1.5 | 1.1s | 0.0s (0%)
argh_shared v0.1.5 | 0.6s | 0.4s (59%)
argh v0.1.5 | 0.2s | 0.1s (26%)



Targets: | naga-cli 0.4.0 (bin "naga")
-- | --
Profile: | dev
Fresh units: | 0
Dirty units: | 56
Total units: | 56
Max concurrency: | 12 (jobs=12 ncpu=12)
Build start: | 2021-07-20T14:47:05Z
Total time: | 18.1s
rustc: | rustc 1.55.0-nightly (a85f584ae 2021-06-16)Host: x86_64-unknown-linux-gnuTarget: x86_64-unknown-linux-gnu
Max (global) rustc threads concurrency: | 0

---

Without `argh`:

Targets: | naga-cli 0.4.0 (bin "naga")
-- | --
Profile: | dev
Fresh units: | 0
Dirty units: | 51
Total units: | 51
Max concurrency: | 12 (jobs=12 ncpu=12)
Build start: | 2021-07-20T14:58:40Z
Total time: | 15.6s
rustc: | rustc 1.55.0-nightly (a85f584ae 2021-06-16)Host: x86_64-unknown-linux-gnuTarget: x86_64-unknown-linux-gnu
Max (global) rustc threads concurrency: | 0

----

In summary, looks like it adds ~1.5 seconds roughly. I didn't run the tests multiple times or anything, though.

**Comment by kvark** [comment]
So it sounds like it's adding 10% of the build time? That's quite noticeable :(

**Comment by zicklag** [comment]
6%, I guess. But for a project that only takes 18 seconds to compile anyway, I would gladly pay 2 seconds of my life for much better documentation and CLI checking. It'll save me a lot more time to wait for it to compile, and the maintenance time documenting the usage, then it will cost waiting for builds.

**Comment by zicklag** [comment]
Timings for release builds:

**With:**

Targets: | naga-cli 0.4.0 (bin "naga")
-- | --
Total time: | 25.3s

**Without:**

Targets: | naga-cli 0.4.0 (bin "naga")
-- | --
Total time: | 23.6s


That's a 1.7s ( 7.2% ) increase in build time on a clean build.

---

I opened #1116 with the argument parser update and switch panics to proper error handling. If you think the compile times are worth it I can just pull `argh` out of that PR I'll add a more extensive error message and keep the better error handling.
